ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/jca/TransactionImporterImple.java

```diff
@@ -211,17 +233,35 @@ public void removeImportedTransaction(Xid xid) throws XAException
 	}
 	
 	public Set<Xid> getInflightXids(String parentNodeName) {
-		Iterator<TransactionImple> iterator = _transactions.values().iterator();
+		Iterator<AtomicReference<TransactionImple>> iterator = _transactions.values().iterator();
 		Set<Xid> toReturn = new HashSet<Xid>();
 		while (iterator.hasNext()) {
-			TransactionImple next = iterator.next();
-			if (next.getParentNodeName().equals(parentNodeName)) {
-				toReturn.add(next.baseXid());
+			AtomicReference<TransactionImple> next = iterator.next();
+			TransactionImple imported = next.get();
+
+			if (imported == null) {
+				// another thread must be in the process of importing this transaction
+				// TODO the solution does not provide a mechanism to wait for the other thread to update the holder
```

Comment:
I added https://github.com/mmusgrov/narayana/pull/2 but it requires public API change